test/integration/reset_sensors.cc

```diff
@@ -0,0 +1,210 @@
+    EXPECT_EQ(0u, this->mockSystem->updateCallCount);
+    EXPECT_EQ(0u, this->mockSystem->postUpdateCallCount);
+  }
+
+  // Run so that things will happen in the world
+  // In this case, the box should fall some
+  server.Run(true, 100, false);
+  {
+    ASSERT_NE(nullptr, ecm);
+    auto entity = ecm->EntityByComponents(components::Name("box"));
+    ASSERT_NE(kNullEntity, entity);
+    auto poseComp = ecm->Component<components::Pose>(entity);
+    ASSERT_NE(nullptr, poseComp);
+    EXPECT_LT(poseComp->Data().Z(), 5.0f);
+
+    EXPECT_EQ(1u, this->mockSystem->configureCallCount);
+    EXPECT_EQ(0u, this->mockSystem->resetCallCount);
+    EXPECT_EQ(100u, this->mockSystem->preUpdateCallCount);
+    EXPECT_EQ(100u, this->mockSystem->updateCallCount);
+    EXPECT_EQ(100u, this->mockSystem->postUpdateCallCount);
+  }
+
+  // Validate update info in the reset
+  this->mockSystem->resetCallback =
+    [](const gazebo::UpdateInfo &_info,
+       gazebo::EntityComponentManager &)
+    {
+      EXPECT_EQ(0u, _info.iterations);
+      EXPECT_EQ(std::chrono::steady_clock::duration{0}, _info.simTime);
+    };
+
+  // Send command to reset to initial state
+  worldReset();
+
+  // It takes two iterations for this to propage, 
+  // the first is for the message to be recieved and internal state setup
+  server.Run(true, 1, false);
+  EXPECT_EQ(1u, this->mockSystem->configureCallCount);
+  EXPECT_EQ(0u, this->mockSystem->resetCallCount);
+  EXPECT_EQ(101u, this->mockSystem->preUpdateCallCount);
+  EXPECT_EQ(101u, this->mockSystem->updateCallCount);
+  EXPECT_EQ(101u, this->mockSystem->postUpdateCallCount);
+
+  // The second iteration is where the reset actually occurs.
+  server.Run(true, 1, false);
+  {
+    ASSERT_NE(nullptr, ecm);
+    auto entity = ecm->EntityByComponents(components::Name("box"));
+    ASSERT_NE(kNullEntity, entity);
+    auto poseComp = ecm->Component<components::Pose>(entity);
+    ASSERT_NE(nullptr, poseComp);
+    EXPECT_FLOAT_EQ(poseComp->Data().Z(), 5.0f);
+
+    EXPECT_EQ(1u, this->mockSystem->configureCallCount);
+    EXPECT_EQ(1u, this->mockSystem->resetCallCount);
+
+    // These should not increment, because only reset is called
+    EXPECT_EQ(101u, this->mockSystem->preUpdateCallCount);
+    EXPECT_EQ(101u, this->mockSystem->updateCallCount);
+    EXPECT_EQ(101u, this->mockSystem->postUpdateCallCount);
+  }
+
+  server.Run(true, 100, false);
+  {
+    ASSERT_NE(nullptr, ecm);
+    auto entity = ecm->EntityByComponents(components::Name("box"));
+    ASSERT_NE(kNullEntity, entity);
+    auto poseComp = ecm->Component<components::Pose>(entity);
+    ASSERT_NE(nullptr, poseComp);
+    EXPECT_LT(poseComp->Data().Z(), 5.0f);
+
+    EXPECT_EQ(1u, this->mockSystem->configureCallCount);
+    EXPECT_EQ(1u, this->mockSystem->resetCallCount);
+    EXPECT_EQ(201u, this->mockSystem->preUpdateCallCount);
+    EXPECT_EQ(201u, this->mockSystem->updateCallCount);
+    EXPECT_EQ(201u, this->mockSystem->postUpdateCallCount);
+  }
```

Comment:
Would it be possible to test the output of the sensors to see that they still publish the correct data after Reset?

Reply:
[4770dc7](https://github.com/gazebosim/gz-sim/pull/1355/commits/4770dc7fa5d1205dbfeff84a9c5d3912ae52c44b)